lib/features/push_notification/data/datasource_impl/remote_web_socket_datasource_impl.dart

```diff
@@ -0,0 +1,93 @@
+import 'dart:convert';
+
+import 'package:core/utils/app_logger.dart';
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:jmap_dart_client/jmap/core/capability/capability_identifier.dart';
+import 'package:jmap_dart_client/jmap/core/capability/web_socket_ticket_capability.dart';
+import 'package:jmap_dart_client/jmap/core/capability/websocket_capability.dart';
+import 'package:jmap_dart_client/jmap/core/session/session.dart';
+import 'package:model/extensions/session_extension.dart';
+import 'package:rxdart/transformers.dart';
+import 'package:tmail_ui_user/features/push_notification/data/datasource/web_socket_datasource.dart';
+import 'package:tmail_ui_user/features/push_notification/data/network/web_socket_api.dart';
+import 'package:tmail_ui_user/features/push_notification/domain/exceptions/web_socket_exceptions.dart';
+import 'package:tmail_ui_user/main/error/capability_validator.dart';
+import 'package:tmail_ui_user/main/exceptions/exception_thrower.dart';
+import 'package:web_socket_channel/web_socket_channel.dart';
+
+class RemoteWebSocketDatasourceImpl implements WebSocketDatasource {
+  final WebSocketApi _webSocketApi;
+  final ExceptionThrower _exceptionThrower;
+  
+  RemoteWebSocketDatasourceImpl(this._webSocketApi, this._exceptionThrower);
+
+  int _webSocketRetryRemained = 3;
+
+  @override
+  Stream getWebSocketChannel(Session session, AccountId accountId) {
+    return Stream
+      .castFrom(_getWebSocketChannel(session, accountId))
+      .doOnError(_exceptionThrower.throwException);
+  }
+
+  Stream _getWebSocketChannel(Session session, AccountId accountId) async* {
+    try {
+      _verifyWebSocketCapabilities(session, accountId);
+      final webSocketTicket = await _generateWebSocketTicket(session, accountId);
+      final webSocketUri = _getWebSocketUri(session, accountId);
+      
+      final webSocketChannel = WebSocketChannel.connect(
+        Uri.parse('$webSocketUri?ticket=$webSocketTicket'),
+        protocols: ['jmap']);
+      await webSocketChannel.ready;
+      webSocketChannel.sink.add(jsonEncode({"@type": "WebSocketPushEnable"}));
```

Comment:
we should add enum for `@type`?

Reply:
Not available anymore. Please review on latest code instead of old commits.